controller/ca/controller.go

```diff
@@ -172,3 +181,15 @@ func (c *CertificateController) handlePodAdd(obj interface{}) {
 func (c *CertificateController) handlePodUpdate(oldObj, newObj interface{}) {
 	c.handlePodAdd(newObj)
 }
+
+func (c *CertificateController) handleMWCAdd(obj interface{}) {
+	mwc := obj.(*v1beta1.MutatingWebhookConfiguration)
+	log.Debugf("enqueuing secret write for mutating webhook configuration %q", mwc.ObjectMeta.Name)
+	for _, webhook := range mwc.Webhooks {
+		c.queue.Add(fmt.Sprintf("%s.%s.%s", webhook.ClientConfig.Service.Name, pkgK8s.Service, webhook.ClientConfig.Service.Namespace))
+	}
+}
```

Comment:
Should this callback check the name of the MWC before doing anything, since we're watching all MWCs in the cluster, but we only care about the "proxy-injector.linkerd.io" MWC? Something like:

```
func (c *CertificateController) handleMWCAdd(obj interface{}) {
	mwc := obj.(*v1beta1.MutatingWebhookConfiguration)
	if mwc.Name == "proxy-injector.linkerd.io" {
		log.Debugf("enqueuing secret write for mutating webhook configuration %q", mwc.ObjectMeta.Name)
		for _, webhook := range mwc.Webhooks {
			c.queue.Add(fmt.Sprintf("%s.%s.%s", webhook.ClientConfig.Service.Name, pkgK8s.Service, webhook.ClientConfig.Service.Namespace))
		}
	}
}
```

But if we do need to do something like that, the name of the MWC should probably be provided as a flag to the `ca` service on startup.

Reply:
I think you are spot on (again). It does look like any future webhooks can all be defined in the same MWC. The [`webhooks`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#webhook-v1beta1-admissionregistration-k8s-io) spec is an array of webhook configurations, each with its own rules, client config, namespace filter etc. Hence, we don't need to the watch.

Updating the `ca` service with an endpoint to generate a secret on-demand sounds reasonable.But how are we going to establish the trust between the `ca` service and the proxy-injector (which on start-up, doesn't have any TLS identity yet). It also implies that other pods can call that endpoint to generate TLS certs signed by the LD2 CA, right? How are we going to establish the trust between the `ca` service and these callers?

Reply:
Those are all great points about the endpoint approach. I think it makes the most sense to stick with the watch approach that you've implemented, but gate it behind a flag as I suggest in my other review, so that we only establish the watch if we're also running the admission controller.